flatten_obj.py

```diff
@@ -0,0 +1,347 @@
+        chain renaming {default: 1}
+
+    chain_map = An attribute name for the 'stored' scratch object. If
+        specified, `stored.<chain_map>` will be populated with a dictionary
+        mapping the new chain names to a tuple giving the originated object,
+        state, and chainID. {default: ""}
+
+NOTES
+
+    Like the select command, if name is omitted then the default object name
+    ("flat") is used as the name argument.
+
+    Chain renaming is tricky. PDB files originally limited chains to single
+    letter identifiers containing [A-Za-z0-9]. When this was found to be
+    limiting, multi-letter chains (ideally < 4 chars) were allowed. This is
+    supported as of PyMOL 1.7. Earlier versions do not accept rename=2, and
+    will raise an exception when flattening a structure with more than 62
+    chains.
+
+EXAMPLES
+
+    flatten_obj flat, nmrObj
+    flatten_obj ( obj1 or obj2 )
+
+PYMOL API
+
+    cmd.flatten_obj(string name, string selection)
+
+SEE ALSO
+
+    split_states
+
+    """
+
+    # arguments
+
+    # Single argument; treat as selection
+    if name and not selection:
+        selection = name
+        name = ""
+    # default name and selection
+    if not name:
+        name = "flat"
+    if not selection:
+        selection = "(all)"
+
+    state = int(state)
+    rename = int(rename)
+    quiet = int(quiet)
+
+    # Wrap in extra parantheses for get_object_list
+    selection = "( %s )" % selection
+
+    if rename == 0:
+        chainSet = DefaultChainSet()
+    elif rename == 1:
+        chainSet = SequentialChainSet()
+    elif rename == 2:
+        chainSet = LongChainSet()
+    else:
+        raise ValueError("Unrecognized rename option (Valid: 0,1,2)")
+
+    metaprefix = "temp" #TODO unique prefix
+
+    # store original value of retain_order, which causes weird interleaving of
+    # structures if enabled.
+    retain_order = cmd.get("retain_order")
+    try:
+        cmd.set("retain_order",0)
+
+        # create new object for each state
+        for obj in cmd.get_object_list(selection):
+
+            if state <= 0:
+                # all states
+                prefix = "%s_%s_"%(metaprefix,obj)
+                cmd.split_states(obj,prefix=prefix)
+            else:
+                prefix = "%s_%s_%04d"%(metaprefix,obj,state)
+                cmd.create(prefix, obj, state, 1)
+
+        # renumber all states
+        statere = re.compile("^%s_(.*)_(\d+)$" % metaprefix) # matches split object names
+
+        warn_lowercase = False
+
+        # Iterate over all objects with metaprefix
+        try:
+            for obj in cmd.get_object_list("(%s_*)"%(metaprefix) ):
+                m = statere.match(obj)
+                if m is None:
+                    print("Failed to match object %s" %obj)
+                    continue
+                origobj = m.group(1)
+                statenum = int(m.group(2))
+
+                chains = set([atom.chain for atom in cmd.get_model(obj).atom])
```

Comment:
cmd.get_chains() should be much more efficient, cmd.get_model() is a lot of overhead